File: lib/ts/src/internal/types/VSelect.ts
import { VTypeDeclaration } from './VTypeDeclaration';
import { ValidationFailure } from '../validation/ValidationFailure';
import { VType } from './VType';
import { validateSelect } from '../validation/ValidateSelect';
import { GenerateContext } from '../generation/GenerateContext';
import { generateRandomSelect } from '../generation/GenerateRandomSelect';
import { ValidateContext } from '../validation/ValidateContext';

export const select: string = 'Object';

export class VSelect implements VType {
    possibleSelects: Record<string, any> = {};

    getTypeParameterCount(): number {
        return 0;
    }

    validate(givenObj: any, typeParameters: VTypeDeclaration[], ctx: ValidateContext): ValidationFailure[] {
        return validateSelect(givenObj, this.possibleSelects, ctx);
    }

    generateRandomValue(
        blueprintValue: any,
        useBlueprintValue: boolean,
        typeParameters: VTypeDeclaration[],
        ctx: GenerateContext,
    ): any {
        return generateRandomSelect(this.possibleSelects, ctx);
    }

    getName(): string {
        return select;
    }
}

File: lib/java/src/main/java/io/github/msgpact/internal/types/VSelect.java
package io.github.msgpact.internal.types;

import static io.github.msgpact.internal.generation.GenerateRandomSelect.generateRandomSelect;
import static io.github.msgpact.internal.validation.ValidateSelect.validateSelect;

import java.util.HashMap;
import java.util.List;
import java.util.Map;

import io.github.msgpact.internal.generation.GenerateContext;
import io.github.msgpact.internal.validation.ValidateContext;
import io.github.msgpact.internal.validation.ValidationFailure;

public class VSelect implements VType {

    public static final String _SELECT = "Object";

    public final Map<String, Object> possibleSelects = new HashMap<>();

    @Override
    public int getTypeParameterCount() {
        return 0;
    }

    @Override
    public List<ValidationFailure> validate(Object givenObj,
            List<VTypeDeclaration> typeParameters, ValidateContext ctx) {
        return validateSelect(givenObj, this.possibleSelects, ctx);
    }

    @Override
    public Object generateRandomValue(Object blueprintValue, boolean useBlueprintValue,
            List<VTypeDeclaration> typeParameters, GenerateContext ctx) {
        return generateRandomSelect(this.possibleSelects, ctx);
    }

    @Override
    public String getName() {
        return _SELECT;
    }

}
